app/src/main/java/com/github/h3lp3rs/h3lp/forum/implementation/CachedForum.kt

```diff
@@ -0,0 +1,178 @@
+                    }
+                }
+            }
+            // Fetch cache for posts if not found -> If fail we only get an empty list
+            if (cachedList.isEmpty()) getAllAux() else cachedList
+        }
+    }
+
+    private fun getAllAux(): List<CategoryPosts> {
+        if (path.isEmpty()) {
+            // In case we are in the root forum, get the CategoryPosts from all categories
+            val ls = mutableListOf<CategoryPosts>()
+            for (category in ForumCategory.values()) {
+                val categoryForum = CachedForum(forum.child(category.name))
+                ls.addAll(categoryForum.getAllAux())
+            }
+
+            return ls
+        } else {
+            // In case we are in a category, we only have a single CategoryPost to return (thus
+            // we return a singleton list)
+            val allPosts = mutableListOf<ForumPost>()
+            val entry = getCacheEntry(emptyList()).posts
+            for (post in entry) {
+                val replies = getCacheEntry(listOf(post.repliesKey)).posts
+                allPosts.add(ForumPost(this, post, replies))
+            }
+
+            return listOf(Pair(path, allPosts))
+        }
+    }
+
+    private fun updateCacheWithPost(relativePath: Path, postData: ForumPostData) {
+        // Copy for mutability issues
+        val posts = ArrayList(getCacheEntry(relativePath).posts)
+
+        // Look for post with same key
+        for (i in 0 until posts.size) {
+            val p = posts[i]
+            if (p.key == postData.key) {
+                posts[i] = postData
+                setCacheEntry(relativePath, CacheEntry(posts))
+                return
+            }
+        }
+
+        // First occurrence of key
+        posts.add(postData)
+        setCacheEntry(relativePath, CacheEntry(posts))
+    }
+
+    private fun getCacheEntry(relativePath: Path): CacheEntry {
+        val cachePath = (path + relativePath).joinToString(separator = "/")
+
+        // Never null due to non-null default parameter
+        return cache.getObjectOrDefault(
+            cachePath,
+            CacheEntry::class.java, CacheEntry(ArrayList())
+        )!!
+    }
+
+    private fun setCacheEntry(relativePath: Path, entry: CacheEntry) {
+        val cachePath = (path + relativePath).joinToString(separator = "/")
+        cache.setObject(cachePath, CacheEntry::class.java, entry)
+    }
+
+    override fun listenToAll(action: (ForumPostData) -> Unit) {
+        val newWrappedAction: (ForumPostData) -> Unit = {
+            // This is a very hacky but easy way to know from the data, the path to its forum pointer
+            // We should change the listener design to make it cleaner, but for now it works fine
+            // Impossible to listen for replies, would need to heavy design changes
+            val path =
+                listOf(it.category.name) + if (it.isPost) emptyList() else listOf(it.repliesKey)

```

Comment:
```suggestion
                listOf(it.category.name) + (if (it.isPost) emptyList() else listOf(it.repliesKey))

```

Just for more readability